In [13]:
import pandas as pd
import numpy as np
from scipy import stats

Empezamos una nueva lección, y este es el momento de poner en práctica los conocimientos adquiridos del Estadística Descriptiva básica. En este caso vamos a darle un giro de tuerca a la forma en la que se ha resuelto los ejercicios de la Pair de hasta ahora! Para ello os vamos a pedir que por vuestra cuenta busqueis un dataset que os guste, por ejemplo de Kaggle o de donde queráis extraerlo. De esta forma vamos a estudiar los diferentes estadísiticos que contiene en dataset seleccionado Para todas los ejercicios de pair programming de Estadística I, II y III usaremos el dataframe que se tome para el desarrollo de este ejercicio de pair programming.

EJERCICIO 1: Buscar un conjunto de datos a analizar: 
Se recomienda que el conjunto de datos a analizar tenga variables numéricas y categóricas. Ya que también se analizarán.

In [14]:
df = pd.read_csv('insurance.csv')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [15]:
df.shape


(1338, 7)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


EJERCICIO 2: Extraer los siguentes estadísiticos del conjunto de datos para varias de las variables disponibles:
- Media
- Desviación con respecto de la media
- Varianza
- Desviación típica
- Moda
- Mediana
- Media ponderada
- Robustez
- Coeficiente de variación

In [17]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

In [18]:
media_edades = np.round(np.mean(df['age']), decimals=2)
print("la media de las edades es:", media_edades)
media_hijos = np.round(np.mean(df['children']), decimals=2)
print("La media de los hijos es:", media_hijos)
media_imc = np.round(np.mean(df['bmi']), decimals=2)
print("La media del IMC es:", media_imc)
media_costos = np.round(np.mean(df['charges']), decimals=2)
print("La media de los costos es:", media_costos)

la media de las edades es: 39.21
La media de los hijos es: 1.09
La media del IMC es: 30.66
La media de los costos es: 13270.42


In [19]:
variacion_media_edades = np.abs(df["age"]-media_edades)
print("la variación de la media de las edades es:", variacion_media_edades)
variacion_media_hijos = np.abs(df['children']-media_hijos)
print("La variación de la media de los hijos es:", variacion_media_hijos)
variacion_media_imc = np.abs(df['bmi']- media_imc)
print("La variación de la media del IMC es:", variacion_media_imc)
variacion_media_costos = np.abs(df['charges']- media_costos)
print("La variación de la media de los costos es:", variacion_media_costos)

la variación de la media de las edades es: 0       20.21
1       21.21
2       11.21
3        6.21
4        7.21
        ...  
1333    10.79
1334    21.21
1335    21.21
1336    18.21
1337    21.79
Name: age, Length: 1338, dtype: float64
La variación de la media de los hijos es: 0       1.09
1       0.09
2       1.91
3       1.09
4       1.09
        ... 
1333    1.91
1334    1.09
1335    1.09
1336    1.09
1337    1.09
Name: children, Length: 1338, dtype: float64
La variación de la media del IMC es: 0       2.760
1       3.110
2       2.340
3       7.955
4       1.780
        ...  
1333    0.310
1334    1.260
1335    6.190
1336    4.860
1337    1.590
Name: bmi, Length: 1338, dtype: float64
La variación de la media de los costos es: 0        3614.50400
1       11544.86770
2        8820.95800
3        8714.05061
4        9403.56480
           ...     
1333     2669.87170
1334    11064.43920
1335    11640.58650
1336    11262.47500
1337    15870.94030
Name: charges, Length: 1338, dtype: flo

In [20]:
varianza_edades = np.round(np.var(df['age']), decimals= 2)
print("la varianza de las edades es:", varianza_edades)
varianza_hijos = np.round(np.var(df['children']), decimals=2)
print("La varianza de los hijos es:", varianza_hijos)
varianza_imc = np.round(np.var(df['bmi']), decimals=2)
print("La varianza del IMC es:", varianza_imc)
varianza_costos = np.round(np.var(df['charges']), decimals=2)
print("La varianza de los costos es:", varianza_costos)

la varianza de las edades es: 197.25
La varianza de los hijos es: 1.45
La varianza del IMC es: 37.16
La varianza de los costos es: 146542766.49


In [21]:
desv_tip_edades = np.round(np.std(df["age"]), decimals=2)
print("La desviación típica de la edad es:", desv_tip_edades)
desv_tip_hijos = np.round(np.std(df["children"]), decimals=2)
print("La desviación típica de los hijos es:", desv_tip_hijos)
desv_tip_imc = np.round(np.std(df["bmi"]), decimals=2)
print("La desviación típica del IMC es:", desv_tip_imc)
desv_tip_costos = np.round(np.std(df["charges"]), decimals=2)
print("La desviación típica de los costos es:", desv_tip_costos)


La desviación típica de la edad es: 14.04
La desviación típica de los hijos es: 1.21
La desviación típica del IMC es: 6.1
La desviación típica de los costos es: 12105.48


In [22]:
df.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')

In [23]:
moda_df, frecuencia_df = stats.mode(df)
print(moda_df)
print(frecuencia_df)


[[18 'male' 32.3 0 'no' 'southeast' 1639.5631]]
[[  69  676   13  574 1064  364    2]]


In [24]:
print(f"la moda de la edad es: {moda_df[0][0]} y su frecuencia: {frecuencia_df[0][0]}")
print(f"la moda del sexo es: {moda_df[0][1]} y su frecuencia: {frecuencia_df[0][1]}")
print(f"la moda del IMC es: {moda_df[0][2]} y su frecuencia: {frecuencia_df[0][2]}")
print(f"la moda de los niños es: {moda_df[0][3]} y su frecuencia: {frecuencia_df[0][3]}")
print(f"la moda de los fumadores es: {moda_df[0][4]} y su frecuencia: {frecuencia_df[0][4]}")
print(f"la moda de la region es: {moda_df[0][5]} y su frecuencia: {frecuencia_df[0][5]}")
print(f"la moda de los costos es: {moda_df[0][6]} y su frecuencia: {frecuencia_df[0][6]}")


la moda de la edad es: 18 y su frecuencia: 69
la moda del sexo es: male y su frecuencia: 676
la moda del IMC es: 32.3 y su frecuencia: 13
la moda de los niños es: 0 y su frecuencia: 574
la moda de los fumadores es: no y su frecuencia: 1064
la moda de la region es: southeast y su frecuencia: 364
la moda de los costos es: 1639.5631 y su frecuencia: 2


In [25]:
mediana_edad=np.median(df["age"])
mediana_bmi=np.median(df["bmi"])
mediana_children=np.median(df["children"])
mediana_charges=np.median(df["charges"])
print(f"La mediana de la columna edad es: {mediana_edad}")
print(f"La mediana de la columna bmi es: {mediana_bmi}")
print(f"La mediana de la columna children es: {mediana_children}")
print(f"La mediana de la columna charges es: {mediana_charges}")

La mediana de la columna edad es: 39.0
La mediana de la columna bmi es: 30.4
La mediana de la columna children es: 1.0
La mediana de la columna charges es: 9382.033


In [26]:
print(np.average(df["age"]))
print(np.average(df["bmi"]))
print(np.average(df["children"]))
print(np.average(df["charges"]))

39.20702541106129
30.66339686098655
1.0949177877429
13270.422265141257


In [27]:
w = np.ones(1338, dtype=np.complex64)
avg_edad=np.round(np.average(df["age"], weights=w), decimals=2)
avg_bmi=np.round(np.average(df["bmi"], weights=w), decimals=2)
avg_children=np.round(np.average(df["children"], weights=w), decimals=2)
avg_charges=np.round(np.average(df["charges"], weights=w), decimals=2)

In [30]:
print(f"La media ponderada de la columna edad es: {avg_edad}")
print(f"La media ponderada de la columna bmi es: {avg_bmi}")
print(f"La media ponderada de la columna children es: {avg_children}")
print(f"La media ponderada de la columna charges es: {avg_charges}")

print(f'El coeficiente de variación de la edad es: {desv_tip_edades/media_edades}')
print(f'El coeficiente de variación del bmi: {desv_tip_imc/media_imc}')
print(f'EL coeficiente de variación de los hijos es: {desv_tip_hijos/media_hijos}')
print(f'EL coeficiente de variación de los costes es: {desv_tip_costos/media_costos}')

La media ponderada de la columna edad es: (39.21+0j)
La media ponderada de la columna bmi es: (30.66+0j)
La media ponderada de la columna children es: (1.09+0j)
La media ponderada de la columna charges es: (13270.42+0j)
El coeficiente de variación de la edad es: 0.3580719204284621
El coeficiente de variación del bmi: 0.1989562948467058
EL coeficiente de variación de los hijos es: 1.1100917431192658
EL coeficiente de variación de los costes es: 0.9122152878356525


EJERCICIO 3: Interpretación de los resultados.
Ahora interpreta los resultados obtenidos de los diferentes estadísticos estudiados haciendo hincapié en aspectos interesantes de tus datos y explica lo mejor posible la naturaleza de los mismos.

Según el análisis descriptivo que hemos hecho de los datos podemos concluir que:

- La media de edad de los clientes de seguro médico es de 39 años. 
- La media de hijos es igual a 1.
- Y los gastos en seguro médico son en promedio de 13270 dólares.
- La media y la mediana para la edad, el bmi y la cantidad de hijos son similares, lo cual nos indica que la distribución de nuestros datos es simétrica. 
- El género más común entre los clientes es masculino. 
- La mayoría de los clientes no tienen hijos y son NO fumadores. 
- En general, los datos de nuestras variables tienen un coeficiente de variación bajo. Esto nos indica que la dispersión de nuestros datos es baja. 